boot/bootstate20.go

```diff
@@ -21,329 +21,518 @@ package boot
+// kernel snaps on UC20. It is used for setNext() and markSuccessful() - though
+// note that for markSuccessful() a different bootStateUpdate implementation is
+// returned, see bootState20MarkSuccessful
+type bootState20Kernel struct {
+	// the bootloader to manipulate kernel assets
+	ebl bootloader.ExtractedRunKernelImageBootloader
+
+	// the kernel_status variable currently in the bootloader, initialized by
+	// setupBootloader()
+	kernelStatus string
 
-	// the trying snap is the one given by the try-kernel.efi symlink, while
-	// the known-good snap is the one given by the kernel.efi symlink
-	// we can determine if we are trying the try snap by the presence of
-	// 1. a try-kernel.efi symlink
-	// 2. kernel_status == trying
+	// the kernel_status value to be written in commit()
+	commitKernelStatus string
 
+	// the kernel snap that was tried for markSuccessful()
+	triedKernelSnap snap.PlaceInfo
+
+	// the kernel snap to try for setNext()
+	tryKernelSnap snap.PlaceInfo
+}
+
+func (ks20 *bootState20Kernel) loadBootloader() error {
+	// don't setup multiple times
+	if ks20.ebl != nil {
+		return nil
+	}
+	// find the bootloader and ensure it's an extracted run kernel image
+	// bootloader
 	bl, err := bootloader.Find("", nil)
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot find any bootloader: %v", err)
+		return err
 	}
-
 	ebl, ok := bl.(bootloader.ExtractedRunKernelImageBootloader)
 	if !ok {
-		return nil, nil, false, fmt.Errorf("cannot use %s bootloader: does not support extracted run kernel images", ebl.Name())
+		return fmt.Errorf("cannot use %s bootloader: does not support extracted run kernel images", bl.Name())
+	}
+
+	ks20.ebl = ebl
+
+	// also get the kernel_status
+	m, err := ebl.GetBootVars("kernel_status")
+	if err != nil {
+		return err
+	}
+
+	ks20.kernelStatus = m["kernel_status"]
+
+	return nil
+}
+
+func (ks20 *bootState20Kernel) revisions() (snap.PlaceInfo, snap.PlaceInfo, string, error) {
+	var bootSn, tryBootSn snap.PlaceInfo
+	err := ks20.loadBootloader()
+	if err != nil {
+		return nil, nil, "", err
 	}
 
 	// get the kernel for this bootloader
-	bootSnap, err = ebl.Kernel()
+	bootSn, err = ks20.ebl.Kernel()
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot identify kernel snap with bootloader %s: %v", ebl.Name(), err)
+		return nil, nil, "", fmt.Errorf("cannot identify kernel snap with bootloader %s: %v", ks20.ebl.Name(), err)
 	}
 
-	tryKernel, tryKernelExists, err := ebl.TryKernel()
+	tryKernel, tryKernelExists, err := ks20.ebl.TryKernel()
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot identify try kernel snap with bootloader %s: %v", ebl.Name(), err)
+		return nil, nil, "", fmt.Errorf("cannot identify try kernel snap with bootloader %s: %v", ks20.ebl.Name(), err)
 	}
 
 	if tryKernelExists {
-		tryBootSnap = tryKernel
+		tryBootSn = tryKernel
 	}
 
-	m, err := bl.GetBootVars("kernel_status")
+	m, err := ks20.ebl.GetBootVars("kernel_status")
```

Comment:
we don't need this bit again, no?  we can just use ks20.kernelStatus

Reply:
ah yes kernelStatus is set in loadBootloader, you're right